boards: arm: arduino_opta_m4: adds initial support to Arduino OPTA PLC on the M4 core (#64101)
zephyrproject-rtos/zephyr: boards/arm/arduino_opta_m4/doc/index.rst

.. _arduino_opta_m4_board:

Arduino OPTA M4-Core
#######################

Overview
********

Opta™ is a secure micro Programmable Logic Controller (PLC)
with Industrial Internet of Things (IoT) capabilities.

Developed in partnership with Finder®, this device supports both the Arduino
programming language and standard IEC-61131-3 PLC programming languages,
such as Ladder Diagram (LD), Sequential Function Chart (SFC),
Function Block Diagram (FBD), Structured Text (ST), and Instruction List (IL),
making it an ideal device for automation engineers.

For Zephyr RTOS, only the M4 is supported for now, making the M7 to run the PLC
tasks while the M4 core under Zephyr acts as a coprocessor.

Additionally, the device features:

- Ethernet compliant with IEEE802.3-2002

> - 256-Mbit SDRAM

Reviewer: Replace with `16MB QSPI Flash`

- 4 x green color status LEDs

> - 1 x user and reset push-button

Reviewer: Reset button is separate and accessible through a pinhole

> - 10 x analog inputs

Reviewer: 8 x analog inputs

.. _arduino_opta_m4_board:

Arduino OPTA M4-Core
#######################

Overview
********

> Opta™ is a secure micro Programmable Logic Controller (PLC)

Reviewer 2: ```suggestion
The Arduino Opta® is a secure micro Programmable Logic Controller (PLC)
```

with Industrial Internet of Things (IoT) capabilities.

Developed in partnership with Finder®, this device supports both the Arduino
programming language and standard IEC-61131-3 PLC programming languages,
such as Ladder Diagram (LD), Sequential Function Chart (SFC),
Function Block Diagram (FBD), Structured Text (ST), and Instruction List (IL),
making it an ideal device for automation engineers.

> For Zephyr RTOS, only the M4 is supported for now, making the M7 to run the PLC

Reviewer 2: ```suggestion
For Zephyr RTOS, only the M4 is supported for now, making the M7 run the PLC
```

| NVIC      | on-chip    | nested vector interrupt controller  |
+-----------+------------+-------------------------------------+
| PINMUX    | on-chip    | pinmux                              |
+-----------+------------+-------------------------------------+
| GPIO      | on-chip    | gpio                                |
+-----------+------------+-------------------------------------+
| FLASH     | on-chip    | flash memory                        |
+-----------+------------+-------------------------------------+
| RNG       | on-chip    | True Random number generator        |
+-----------+------------+-------------------------------------+
| IPM       | on-chip    | virtual mailbox based on HSEM       |
+-----------+------------+-------------------------------------+

Other hardware features are not yet supported on Zephyr porting.

The default configuration per core can be found in the defconfig files:
``boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig``

Pin Mapping
===========

ARDUINO OPTA M4 has access to the 9 GPIO controllers. These controllers are responsible for pin muxing,
input/output, pull-up, etc.

For mode details please refer to `ARDUINO-OPTA website`_.

> Default Zephyr Peripheral Mapping:
> ----------------------------------

Reviewer 2: ```suggestion
Default Zephyr Peripheral Mapping
---------------------------------
```

- Status LED1 : PI0
- Status LED2 : PI1
- Status LED3 : PI3
- Status LED4 : PH15
- User button : PE4

System Clock
============

The STM32H747I System Clock can be driven by an internal or external oscillator,
as well as by the main PLL clock. By default, the CPU2 (Cortex-M4) System clock

> is driven at 240MHz. PLL clock is feed by a 25MHz high speed external clock.

Reviewer 2: ```suggestion
is driven at 240MHz. PLL clock is fed by a 25MHz high speed external clock.
```

Resources sharing
=================

The dual core nature of STM32H747 SoC requires sharing HW resources between the
two cores. This is done in 3 ways:

- **Compilation**: Clock configuration is only accessible to M7 core. M4 core only
  has access to bus clock activation and deactivation.
- **Static pre-compilation assignment**: Peripherals such as a UART are assigned in
  devicetree before compilation. The user must ensure peripherals are not assigned
  to both cores at the same time.
- **Run time protection**: Interrupt-controller and GPIO configurations could be
  accessed by both cores at run time. Accesses are protected by a hardware semaphore
  to avoid potential concurrent access issues.

Programming and Debugging
*************************

> Applications for the ``arduino_opta_m4`` use the regular Zephyr build commands

Reviewer 2: ```suggestion
Applications for the ``arduino_opta_m4`` use the regular Zephyr build commands.
```

> Applications for the ``arduino_opta_m4`` uses the regular Zephyr build commands

Reviewer 2: ```suggestion
Applications for the ``arduino_opta_m4`` use the regular Zephyr build commands
```

See :ref:`build_an_application` for more information about application builds.

Flashing
========

Flashing operation will depend on the target to be flashed and the SoC

> option bytes configuration. The OPTA has an DFU capable bootloader which

Reviewer 2: ```suggestion
option bytes configuration. The OPTA has a DFU capable bootloader which
```

> can be accessed by connecting the device to the USB, and then pressing
> short time the RESET two times, the RESET-LED on the board will fade
> indicating the board is in bootloader.

Reviewer 2: ```suggestion
can be accessed by connecting the device to the USB, and then pressing
the RESET button shortly twice; the RESET-LED on the board will fade
indicating the board is in bootloader mode.
```

For Zephyr RTOS, only the M4 is supported for now, making the M7 to run the PLC
tasks while the M4 core under Zephyr acts as a coprocessor.

Additionally, the device features:

- Ethernet compliant with IEEE802.3-2002
- 256-Mbit SDRAM
- 4 x green color status LEDs
- 1 x user and reset push-button
- 10 x analog inputs
- 4 x isolated relay outputs

.. image:: img/arduino_opta.jpeg
     :align: center
     :alt: ARDUINO-OPTA

More information about the board can be found at the `ARDUINO-OPTA website`_.
More information about STM32H747XIH6 can be found here:

- `STM32H747XI on www.st.com`_
- `STM32H747xx reference manual`_
- `STM32H747xx datasheet`_

Supported Features
==================

The current Zephyr arduino_opta_m4 board configuration supports the following hardware features:

+-----------+------------+-------------------------------------+
| Interface | Controller | Driver/Component                    |
+===========+============+=====================================+
| NVIC      | on-chip    | nested vector interrupt controller  |
+-----------+------------+-------------------------------------+
| UART      | on-chip    | serial port-polling;                |

> |           |            | serial port-interrupt               |

Reviewer 3: drop, not defined in the dts file

+-----------+------------+-------------------------------------+
| PINMUX    | on-chip    | pinmux                              |
+-----------+------------+-------------------------------------+
| GPIO      | on-chip    | gpio                                |
+-----------+------------+-------------------------------------+
| FLASH     | on-chip    | flash memory                        |
+-----------+------------+-------------------------------------+
| RNG       | on-chip    | True Random number generator        |
+-----------+------------+-------------------------------------+
| SPI       | on-chip    | spi                                 |
+-----------+------------+-------------------------------------+

> | IPM       | on-chip    | virtual mailbox based on HSEM       |

Reviewer 3: prob should be dropped as well?

Author: This is something I find ambiguous sometimes, because although both SPI and UART are not in the DTS file, saying they are not supported is not entirely true, since just an overlay fragment makes it to be usable, so put on the boarrd document that other features not mentioned are not supported may discourage users to try to use other features that are actually well supported by the SoC that is not mentioned in its board documentation, so I'm wondering if  we can put on this table two extra columns called supported by SoC, supported ready on the Board.

Anyway, it is just a personal opinion, for now I will remove both UART and SPI :)

+-----------+------------+-------------------------------------+

Other hardware features are not yet supported on Zephyr porting.

The default configuration per core can be found in the defconfig files:
``boards/arm/arduino_opta_m4/arduino_opta_m4_defconfig``

Pin Mapping
===========

ARDUINO OPTA M4 has access to the 9 GPIO controllers. These controllers are responsible for pin muxing,
input/output, pull-up, etc.

For mode details please refer to `ARDUINO-OPTA website`_.

Default Zephyr Peripheral Mapping:
----------------------------------

- Status LD1 : PI0
- Status LD2 : PI1
- Status LD3 : PI3
- Status LD4 : PH15

> - USER : PE4

Reviewer 3: nit, I'd use `Status LED` and `User button`

.. _arduino_opta_m4_board:

Arduino OPTA M4-Core
#######################

Overview
********

Opta™ is a secure micro Programmable Logic Controller (PLC)
with Industrial Internet of Things (IoT) capabilities.

Developed in partnership with Finder®, this device supports both the Arduino
programming language and standard IEC-61131-3 PLC programming languages,
such as Ladder Diagram (LD), Sequential Function Chart (SFC),
Function Block Diagram (FBD), Structured Text (ST), and Instruction List (IL),
making it an ideal device for automation engineers.

For Zephyr RTOS, only the M4 is supported for now, making the M7 to run the PLC
tasks while the M4 core under Zephyr acts as a coprocessor.

Additionally, the device features:

- Ethernet compliant with IEEE802.3-2002

> - USB OTG HS

Reviewer: Still some HW configuration to be tuned:
* USB FS
* 128Mb QUAD-SPI flash

All other info is correct :slightly_smiling_face:

.. _arduino_opta_m4_board:

Arduino OPTA M4-Core
#######################

Overview
********

Opta™ is a secure micro Programmable Logic Controller (PLC)
with Industrial Internet of Things (IoT) capabilities.

Developed in partnership with Finder®, this device supports both the Arduino
programming language and standard IEC-61131-3 PLC programming languages,
such as Ladder Diagram (LD), Sequential Function Chart (SFC),
Function Block Diagram (FBD), Structured Text (ST), and Instruction List (IL),
making it an ideal device for automation engineers.

For Zephyr RTOS, only the M4 is supported for now, making the M7 to run the PLC
tasks while the M4 core under Zephyr acts as a coprocessor.

> Additionally, the device features:

Reviewer: The feature list looks like a copy paste mistake from the EVK :slightly_smiling_face:

two cores. This is done in 3 ways:

- **Compilation**: Clock configuration is only accessible to M7 core. M4 core only
  has access to bus clock activation and deactivation.
- **Static pre-compilation assignment**: Peripherals such as a UART are assigned in
  devicetree before compilation. The user must ensure peripherals are not assigned
  to both cores at the same time.
- **Run time protection**: Interrupt-controller and GPIO configurations could be
  accessed by both cores at run time. Accesses are protected by a hardware semaphore
  to avoid potential concurrent access issues.

Programming and Debugging
*************************

Applications for the ``arduino_opta_m4`` uses the regular Zephyr build commands
See :ref:`build_an_application` for more information about application builds.

Flashing
========

Flashing operation will depend on the target to be flashed and the SoC
option bytes configuration. The OPTA has an DFU capable bootloader which
can be accessed by connecting the device to the USB, and then pressing
short time the RESET two times, the RESET-LED on the board will fade
indicating the board is in bootloader.

By default:

>   - CPU2 (Cortex-M4) boot address is set to 0x81000000 (OB: BOOT_CM4_ADD0)

Reviewer: The PLC binary running on the M7 could grow a bit in the future, maybe it's better to dedicate a bit less flash to the M4? I think 512K would be sensible. Then on the Arduino side we can just call bootM4() https://github.com/arduino/ArduinoCore-mbed/blob/main/variants/OPTA/variant.cpp#L380-L384  with the right address

Author: Yes, I noticed in Arduino IDE it is possible to do flash partitioning in order to leave 512k of flash memory to the M4 instead of equal split.

Author: Moved the flash1 reagion 512K ahead, and left 1.5MB for Arduino and 512K upper for the M4 + Zephyr.

Reviewer: Super, I've seen the change in the dts but the documentation still uses 0x81000000 as default address

can be accessed by connecting the device to the USB, and then pressing
the RESET button shortly twice, the RESET-LED on the board will fade
indicating the board is in bootloader mode.

By default:

  - CPU2 (Cortex-M4) boot address is set to 0x08180000 (OB: BOOT_CM4_ADD0)

Zephyr flash configuration has been set to meet these default settings.

Flashing an application to ARDUINO OPTA M4
------------------------------------------

First, connect the device to your host computer using
the USB port to prepare it for flashing. Then build and flash your application.

Here is an example for the :zephyr:code-sample:`blinky` application on M4 core.

.. zephyr-app-commands::
   :zephyr-app: samples/basic/blinky
   :board: arduino_opta_m4
   :goals: build flash

Starting the application on the ARDUINO OPTA M4
-----------------------------------------------

Make sure the option bytes are set to prevent the M4 from auto-starting, and
that the M7 side starts the M4 at the correct Flash address.

This can be done by selecting in the Arduino IDE's "Tools" / "Flash Split"
menu the "1.5MB M7 + 0.5MB M4" option, and loading a sketch that contains
at least the following code:

>    .. code-block:: cpp
> 
>       #include <RPC.h>
> 
>       void setup() {
>           RPC.begin();
>       }
> 
>       void loop() { }

Reviewer 2: ```suggestion
 .. code-block:: cpp

    #include <RPC.h>

    void setup() {
        RPC.begin();
    }

    void loop() { }
```
